src/main/java/org/spongepowered/common/data/processor/data/ColoredDataProcessor.java

```diff
@@ -0,0 +1,149 @@
+import net.minecraft.init.Items;
+import net.minecraft.item.EnumDyeColor;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemArmor;
+import net.minecraft.item.ItemBlock;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.world.IWorldNameable;
+
+public class ColoredDataProcessor extends AbstractSpongeDataProcessor<ColoredData, ImmutableColoredData> {
+
+    @Override
+    public boolean supports(DataHolder dataHolder) {
+        return dataHolder instanceof EntitySheep || dataHolder instanceof ItemStack || dataHolder instanceof IWorldNameable;
+    }
+
+    @Override
+    public Optional<ColoredData> from(DataHolder dataHolder) {
+        Color color = null;
+        if (dataHolder instanceof EntitySheep) {
+            color = ColorUtil.colorFromDyeColor(((EntitySheep) dataHolder).getFleeceColor());
+        } else if (dataHolder instanceof ItemStack) {
+            final ItemStack stack = (ItemStack) dataHolder;
+            final Item item = stack.getItem();
+            final int meta = stack.getItemDamage();
+            if (item == Items.dye) {
+                // Interpret dyes via "dye damage"
+                color = ColorUtil.colorFromDyeColor(EnumDyeColor.byDyeDamage(meta));
+            } else if (item instanceof ItemBlock) {
+                final Block block = ((ItemBlock) item).getBlock();
+                if (block == Blocks.wool) {
+                    // Interpret wools via metadata
+                    color = ColorUtil.colorFromDyeColor(EnumDyeColor.byMetadata(meta));
+                }
+            } else if (item instanceof ItemArmor) {
+                color = new Color(((ItemArmor) item).getColor(stack));
+            } else if (stack.hasTagCompound()) {
+                final NBTTagCompound tag = stack.getTagCompound();
+                if (tag.hasKey("display", NbtDataUtil.TAG_COMPOUND)) {
+                    final NBTTagCompound display = tag.getCompoundTag("display");
+                    if (display.hasKey("color", NbtDataUtil.TAG_INT)) {
+                        color = new Color(display.getInteger("color"));
+                    }
+                }
+            }
+        }
+        return Optional.ofNullable(color).map(SpongeColoredData::new);
+    }
+
+    @Override
+    public Optional<ColoredData> fill(DataHolder dataHolder, ColoredData manipulator, MergeFunction overlap) {
+        if (supports(dataHolder)) {
+            final ColoredData data = from(dataHolder).orElse(null);
+            final ColoredData newData = checkNotNull(overlap.merge(checkNotNull(manipulator), data));
+            final Color color = newData.color().get();
+            return Optional.of(manipulator.set(Keys.COLOR, color));
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<ColoredData> fill(DataContainer container, ColoredData colorData) {
+        final Color color = DataUtil.getData(container, Keys.COLOR, Color.class);
```

Comment:
Needs to handle the case that if the container doesn't have the key, return `Optional.empty()`.